pkg/querier/worker/scheduler_processor_test.go

```diff
@@ -423,26 +440,82 @@ func TestSchedulerProcessor_ResponseStream(t *testing.T) {
-		}).Return(
+		requestHandler.On("Handle", mock.Anything, mock.Anything).Return(
 			&httpgrpc.HTTPResponse{Code: http.StatusOK, Headers: []*httpgrpc.Header{streamingEnabledHeader},
 				Body: bytes.Repeat([]byte("a"), responseBodySize)},
 			nil,
 		)
 
-		sp.processQueriesOnSingleStream(workerCtx, nil, "127.0.0.1")
+		workerCtx, workerCancel := context.WithCancel(context.Background())
+		go func() {
+			sp.processQueriesOnSingleStream(workerCtx, nil, "127.0.0.1")
+		}()
 
-		assert.Equal(t, 1, int(frontend.queryResultStreamMetadataCalls.Load()))
-		// We expect to receive only part of the body before the stream aborts.
+		assert.Eventually(t, func() bool {
+			return int(frontend.queryResultStreamMetadataCalls.Load()) == 1
+		}, time.Second, 10*time.Millisecond, "expected frontend to be called with response metadata once")
+		assert.Eventually(t, func() bool {
+			return int(frontend.queryResultStreamReturned.Load()) == 1
+		}, time.Second, 10*time.Millisecond, "expected frontend QueryResultStream to have returned once")
+
+		assert.NotNil(t, frontend.responses[queryID].body)
 		assert.Less(t, len(frontend.responses[queryID].body), responseBodySize)
+
+		workerCancel()
+	})
+
+	t.Run("should finish response stream on non-cancellation error", func(t *testing.T) {
+		sp, loopClient, requestHandler, frontend := prepareSchedulerProcessor(t)
+		sp.streamingEnabled = true
+		sp.maxMessageSize = 5 * responseStreamingBodyChunkSizeBytes
+
+		recvCount := atomic.NewInt64(0)
+		frontend.responseStreamStarted = make(chan struct{})
+
+		queryID := uint64(1)
+		loopClient.On("Recv").Return(func() (*schedulerpb.SchedulerToQuerier, error) {
+			switch recvCount.Inc() {
+			case 1:
+				return &schedulerpb.SchedulerToQuerier{
+					QueryID:         queryID,
+					FrontendAddress: frontend.addr,
+					UserID:          "test",
+				}, nil
+			default:
+				<-frontend.responseStreamStarted
+				// Simulate non-cancellation error.
+				return nil, errors.New("something went wrong")
+			}
+		})
+
+		responseBodySize := 4*responseStreamingBodyChunkSizeBytes + 1
+		requestHandler.On("Handle", mock.Anything, mock.Anything).Return(
+			&httpgrpc.HTTPResponse{Code: http.StatusOK, Headers: []*httpgrpc.Header{streamingEnabledHeader},
+				Body: bytes.Repeat([]byte("a"), responseBodySize)},
+			nil,
+		)
+
+		workerCtx, workerCancel := context.WithCancel(context.Background())
+		go func() {
+			sp.processQueriesOnSingleStream(workerCtx, nil, "127.0.0.1")
+		}()
+
+		assert.Eventually(t, func() bool {
+			return int(frontend.queryResultStreamMetadataCalls.Load()) == 1
+		}, time.Second, 10*time.Millisecond, "expected frontend to be called with response metadata once")
+		assert.Eventually(t, func() bool {
+			return int(frontend.queryResultStreamReturned.Load()) == 1
+		}, time.Second, 10*time.Millisecond, "expected frontend QueryResultStream to have returned once")
+
+		assert.Equal(t, len(frontend.responses[queryID].body), responseBodySize)
+
+		workerCancel()
```

Comment:
if the worker is cancelled the query should continue being processed. Previously you cancelled the worker right when it started handling the query. I think you should still cancel it before it has had the time to transmit the response (or even start transmitting it) and then assert that the full response is received.

Reply:
Makes sense, [1860f46](https://github.com/grafana/mimir/pull/7173/commits/1860f4633748ed919ca78d26d2681c5e7eafb273) should cover this.